src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,195 @@
+        }
+        if ((size - 1) == index) {
+            Node<T> node = new Node<>(null, value, null);
+            node.prev = tail.prev;
+            tail.prev = node;
+            node.next = tail;
+            node.prev.next = node;
+            size++;
+        }
+        if (index == 0 && size > 0) {
+            Node<T> node = new Node<>(null, value, null);
+            head.prev = node;
+            node.next = head;
+            head = node;
+            head.prev = null;
+            size++;
+        }
+        if (index == size) {
+            add(value);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
+        checkIndex(index);
+        if (size == 0) {
+            return null;
+        }
+        if (index == 0) {
+            return head.item;
+        }
+        if (index > 0) {
+            Node<T> currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+            return currentNode.item;
+        }
         return null;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = get(index);
+        if (size > 0 && index < (size - 1) && index > 0) {
+            Node<T> node = new Node<>(head,value,tail);
+            for (int i = 1; i < index; i++) {
+                node.prev = node.prev.next;
+            }
+            node.prev.next = node;
+            for (int i = 1; i < size - index - 1; i++) {
+                node.next = node.next.prev;
+            }
+            node.next.prev = node;
+        }
+        if (index == 0 && size > 0) {
+            Node<T> node = new Node<>(null, value, head.next);
+            head.next.prev = node;
+            head = node;
+        }
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        final T removedValue = get(index);
+        size--;
+        if (size == 0) {
+            head = null;
+        }
+        if (index == 0 && size > 0) {
+            head = head.next;
+            head.prev = null;
+        }
+        if (index == (size) && size > 0) {
+            tail = tail.prev;
+            tail.next = null;
+        }
+        if (index > 0 && index < (size)) {
+            Node<T> nodeBeforeIndex = head;
+            for (int i = 1; i < index; i++) {
+                nodeBeforeIndex = nodeBeforeIndex.next;
+            }
+            nodeBeforeIndex.next = nodeBeforeIndex.next.next;
+            Node<T> nodeAfterIndex = tail;
+            for (int i = 1; i < size - index; i++) {
+                nodeAfterIndex = nodeAfterIndex.prev;
+            }
+            nodeAfterIndex.prev = nodeAfterIndex.prev.prev;
+        }
+        return removedValue;
```

Comment:
Please reuse your existing code here.

"Don’t create repeating code.
If the logic of your code repeats - move it to the separate private method. Remember about DRY and KISS principles."